crates/ruff_linter/src/linter.rs

```diff
@@ -268,23 +269,30 @@ pub fn check_path(
     }
 
     // Update fix applicability to account for overrides
-    if !settings.extend_safe_fixes.is_empty() || !settings.extend_unsafe_fixes.is_empty() {
+    if !settings.fix_safety_table.is_empty() {
         for diagnostic in &mut diagnostics {
             if let Some(fix) = diagnostic.fix.take() {
-                // Enforce demotions over promotions so if someone puts a rule in both we are conservative
-                if fix.applicability().is_safe()
-                    && settings
-                        .extend_unsafe_fixes
-                        .contains(diagnostic.kind.rule())
-                {
-                    diagnostic.set_fix(fix.with_applicability(Applicability::Unsafe));
-                } else if fix.applicability().is_unsafe()
-                    && settings.extend_safe_fixes.contains(diagnostic.kind.rule())
-                {
-                    diagnostic.set_fix(fix.with_applicability(Applicability::Safe));
-                } else {
-                    // Retain the existing fix (will be dropped from `.take()` otherwise)
-                    diagnostic.set_fix(fix);
+                match fix.applicability() {
+                    Applicability::Display => {
+                        // If the fix is display-only, we don't need to do anything
+                        // Retain the existing fix (will be dropped from `.take()` otherwise)
+                        diagnostic.set_fix(fix);
+                    }
+                    Applicability::Safe | Applicability::Unsafe => match settings
+                        .fix_safety_table
+                        .resolve_rule(diagnostic.kind.rule())
+                    {
+                        FixSafety::ForcedSafe => {
+                            diagnostic.set_fix(fix.with_applicability(Applicability::Safe));
+                        }
+                        FixSafety::ForcedUnsafe => {
+                            diagnostic.set_fix(fix.with_applicability(Applicability::Unsafe));
+                        }
+                        FixSafety::Default => {
+                            // Retain the existing fix (will be dropped from `.take()` otherwise)
+                            diagnostic.set_fix(fix);
+                        }
+                    },
```

Comment:
This feels kind of awkward, could the `FixSafetyTable` provide a `resolve_applicability(Rule, Applicability) -> Applicability` method instead? Then we could do:

```
diagnostic.set_fix(fix.with_applicability(fix_safety.resolve_applicability(rule, applicability))
```

and push this logic into the safety table